test/lib/jdk/test/lib/net/SimpleHttpServer.java

```diff
@@ -0,0 +1,153 @@
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package jdk.test.lib.net;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.net.InetSocketAddress;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.file.Files;
+import java.nio.file.InvalidPathException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+import com.sun.net.httpserver.Headers;
+import com.sun.net.httpserver.HttpExchange;
+import com.sun.net.httpserver.HttpHandler;
+import com.sun.net.httpserver.HttpServer;
+
+/**
+ * A simple HTTP Server.
+ **/
+public class SimpleHttpServer {
+    private final HttpServer httpServer;
+    private ExecutorService executor;
+    private String address;
+    private final String context;
+    private final String docRoot;
+    private final InetSocketAddress inetSocketAddress;
+
+    public SimpleHttpServer(final InetSocketAddress inetSocketAddress, final String context, final String docRoot)
+            throws IOException {
+        this.inetSocketAddress = inetSocketAddress;
+        this.context = context;
+        this.docRoot = docRoot;
+        httpServer = HttpServer.create();
+    }
+
+    public void start() throws IOException, URISyntaxException {
+        MyHttpHandler handler = new MyHttpHandler(docRoot);
+        httpServer.bind(inetSocketAddress, 0);
+        httpServer.createContext(context, handler);
+        executor = Executors.newCachedThreadPool();
+        httpServer.setExecutor(executor);
+        httpServer.start();
+
+        address = "http:"+URIBuilder.newBuilder().host(httpServer.getAddress().getAddress()).
+                port(httpServer.getAddress().getPort()).build().toString();
+    }
+
+    public void stop() {
+        httpServer.stop(2);
+        executor.shutdown();
+    }
+
+    public String getAddress() {
+        return address;
+    }
+
+    public int getPort() {
+        return httpServer.getAddress().getPort();
+    }
+
+    class MyHttpHandler implements HttpHandler {
+
+        private final String docRoot;
+
+        MyHttpHandler(final String docroot) {
+            docRoot = docroot;
+        }
+
+        public void handle(final HttpExchange t) throws IOException {
+            try (InputStream is = t.getRequestBody()) {
+                is.readAllBytes();
+                Headers rMap = t.getResponseHeaders();
+                try (OutputStream os = t.getResponseBody()) {
+                    URI uri = t.getRequestURI();
+                    String path = uri.getPath();
+                    Path fPath;
+                    try {
+                        fPath = Paths.get(docRoot, path);
```

Comment:
Hi Mahendra,

I am not comfortable with this. You cannot just concatenate a file system path with a URI path in the general case (think windows file system). I would suggest to transform docRoot into a root URI using `Path.of(docRoot).toURI().normalize()` - then you can probably concatenate the string representation of this root URI with the *raw path* extracted from the request URI, normalize that again (to remove potential double slashes) and convert that back into a Path using Path.of(URI). 
Some additional conditions might need to be asserted:
 - `t.getRequestURI().getRawPath()` should either be empty or start with `"/"` (and even that may be a bit lax)
 - the resulting URI you obtain (after normalization and before converting back to Path) should either be equal to the root URI or start with the root URI.
`(rootURI.toString().equals(uri.toString()) || uri.toString().startsWith(rootUri.toString() + "/") )`

best regards,

-- daniel

Reply:
Thanks Daniel. I have implemented the review comments.